File: src/main/java/com/tkr/desk/FeeAccrualEngine.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Tiered fee accrual engine with flat, notional, per-share schedules.
 * Ported from legacy-cpp fee_accrual_engine.cc.
 */
public final class FeeAccrualEngine {

    public enum FeeScheduleKind {
        FLAT_PER_TRADE, TIERED_NOTIONAL, PER_SHARE_MILLI, PERFORMANCE_FEE, MONTHLY_MINIMUM
    }

    public static final class FeeTier {
        public long notionalFloorCents;
        public long notionalCeilingCents;
        public int rateBp;
        public long flatFeeCents;
    }

    public static final class FeeSchedule {
        public FeeScheduleKind kind;
        public int scheduleId;
        public int accountId;
        public List<FeeTier> tiers = new ArrayList<>();
        public int perShareMilliCents;
        public long monthlyMinimumCents;
    }

    public static final class AccrualEntry {
        public int accountId;
        public int scheduleId;
        public long accruedCents;
        public long notionalBasisCents;
        public int tradeCount;
        public int accrualDateYyyymmdd;
    }

    public static final class FeeAccrualEngineConfig {
        public int deskId;
        public int accrualDateYyyymmdd;

        public FeeAccrualEngineConfig(int deskId, int accrualDateYyyymmdd) {
            this.deskId = deskId;
            this.accrualDateYyyymmdd = accrualDateYyyymmdd;
        }
    }

    public static final class FeeAccrualEngineResult {
        public Status status = Status.OK;
        public long totalAccruedCents;
        public int schedulesApplied;
        public List<AccrualEntry> entries = new ArrayList<>();
    }

    private static final class FeeBreakdown {
        long commissionCents;
        long exchangeCents;
        long clearingCents;
        long regulatoryCents;
    }

    private final FeeAccrualEngineConfig config;
    private final List<FeeSchedule> schedules = new ArrayList<>();
    private final List<AccrualEntry> ledger = new ArrayList<>();

    public FeeAccrualEngine() {
        this(new FeeAccrualEngineConfig(0, 0));
    }

    public FeeAccrualEngine(FeeAccrualEngineConfig config) {
        this.config = config != null ? config : new FeeAccrualEngineConfig(0, 0);
        loadDefaultSchedules();
    }

    private static FeeBreakdown computeFeeBreakdown(long notionalCents, int qtyMilli) {
        FeeBreakdown bd = new FeeBreakdown();
        bd.commissionCents = notionalCents * 5 / 10000;
        bd.exchangeCents = notionalCents / 10000;
        bd.clearingCents = (long) qtyMilli * 2 / 1000;
        bd.regulatoryCents = notionalCents * 2 / 10000;
        return bd;
    }

    public void loadDefaultSchedules() {
        schedules.clear();
        FeeSchedule flat = new FeeSchedule();
        flat.kind = FeeScheduleKind.FLAT_PER_TRADE;
        flat.scheduleId = 1;
        FeeTier flatTier = new FeeTier();
        flatTier.notionalFloorCents = 0;
        flatTier.notionalCeilingCents = Long.MAX_VALUE;
        flatTier.flatFeeCents = 250;
        flat.tiers.add(flatTier);
        schedules.add(flat);

        FeeSchedule tiered = new FeeSchedule();
        tiered.kind = FeeScheduleKind.TIERED_NOTIONAL;
        tiered.scheduleId = 2;
        tiered.tiers.add(tier(0, 10_000_000L, 10, 0));
        tiered.tiers.add(tier(10_000_000L, 100_000_000L, 5, 0));
        tiered.tiers.add(tier(100_000_000L, Long.MAX_VALUE, 2, 0));
        schedules.add(tiered);

        FeeSchedule perShare = new FeeSchedule();
        perShare.kind = FeeScheduleKind.PER_SHARE_MILLI;
        perShare.scheduleId = 3;
        perShare.perShareMilliCents = 3;
        schedules.add(perShare);
    }

    private static FeeTier tier(long floor, long ceiling, int rateBp, long flat) {
        FeeTier t = new FeeTier();
        t.notionalFloorCents = floor;
        t.notionalCeilingCents = ceiling;
        t.rateBp = rateBp;
        t.flatFeeCents = flat;
        return t;
    }

    public void registerSchedule(FeeSchedule schedule) {
        if (schedule != null) schedules.add(schedule);
    }

    public void clearSchedules() {
        schedules.clear();
        ledger.clear();
    }

    public long accrueFlatPerTrade(FeeSchedule schedule) {
        return schedule.tiers.isEmpty() ? 0 : schedule.tiers.get(0).flatFeeCents;
    }

    public long accrueTieredNotional(FeeSchedule schedule, long notionalCents) {
        for (FeeTier tier : schedule.tiers) {
            if (notionalCents >= tier.notionalFloorCents && notionalCents < tier.notionalCeilingCents) {
                return notionalCents * tier.rateBp / 10000L + tier.flatFeeCents;
            }
        }
        return 0;
    }

    public long accruePerShareMilli(FeeSchedule schedule, int qtyMilli) {
        return (long) qtyMilli * schedule.perShareMilliCents / 1000L;
    }

    public long accrueMonthlyMinimum(FeeSchedule schedule, long runningTotal) {
        return runningTotal >= schedule.monthlyMinimumCents ? 0 : schedule.monthlyMinimumCents - runningTotal;
    }

    public long computeTradeFee(FeeSchedule schedule, int qtyMilli, long notionalCents) {
        long baseFee = switch (schedule.kind) {
            case FLAT_PER_TRADE -> accrueFlatPerTrade(schedule);
            case TIERED_NOTIONAL -> accrueTieredNotional(schedule, notionalCents);
            case PER_SHARE_MILLI -> accruePerShareMilli(schedule, qtyMilli);
            case PERFORMANCE_FEE -> notionalCents * 2000 / 10000L;
            case MONTHLY_MINIMUM -> schedule.monthlyMinimumCents;
        };
        FeeBreakdown bd = computeFeeBreakdown(notionalCents, qtyMilli);
        return baseFee + bd.exchangeCents + bd.clearingCents + bd.regulatoryCents;
    }

    public void upsertLedger(int accountId, int scheduleId, long feeCents, long notionalCents) {
        for (AccrualEntry entry : ledger) {
            if (entry.accountId == accountId && entry.scheduleId == scheduleId) {
                entry.accruedCents += feeCents;
                entry.notionalBasisCents += notionalCents;
                entry.tradeCount++;
                return;
            }
        }
        AccrualEntry entry = new AccrualEntry();
        entry.accountId = accountId;
        entry.scheduleId = scheduleId;
        entry.accruedCents = feeCents;
        entry.notionalBasisCents = notionalCents;
        entry.tradeCount = 1;
        entry.accrualDateYyyymmdd = config.accrualDateYyyymmdd;
        ledger.add(entry);
    }

    public FeeAccrualEngineResult accrueBatch(BatchWireFrame frame) {
        FeeAccrualEngineResult result = new FeeAccrualEngineResult();
        result.status = Status.OK;
        if (frame == null) return result;
        for (WireBatchRecord rec : frame.records) {
            long notional = (long) rec.qtyMilli * rec.priceTick / 1000L;
            for (FeeSchedule schedule : schedules) {
                if (schedule.accountId != 0 && schedule.accountId != rec.accountId) continue;
                long fee = computeTradeFee(schedule, rec.qtyMilli, notional);
                if (schedule.kind == FeeScheduleKind.MONTHLY_MINIMUM) {
                    long running = 0;
                    for (AccrualEntry entry : ledger) {
                        if (entry.accountId == rec.accountId && entry.scheduleId == schedule.scheduleId) {
                            running = entry.accruedCents;
                            break;
                        }
                    }
                    long minTopUp = accrueMonthlyMinimum(schedule, running + fee);
                    upsertLedger(rec.accountId, schedule.scheduleId, fee + minTopUp, notional);
                    result.totalAccruedCents += fee + minTopUp;
                } else {
                    upsertLedger(rec.accountId, schedule.scheduleId, fee, notional);
                    result.totalAccruedCents += fee;
                }
                result.schedulesApplied++;
            }
        }
        result.entries = new ArrayList<>(ledger);
        result.entries.sort(Comparator.comparingInt((AccrualEntry a) -> a.accountId).thenComparingInt(a -> a.scheduleId));
        return result;
    }

    public List<AccrualEntry> ledgerSnapshot() {
        return new ArrayList<>(ledger);
    }

    public long totalForAccount(int accountId) {
        long sum = 0;
        for (AccrualEntry e : ledger) if (e.accountId == accountId) sum += e.accruedCents;
        return sum;
    }
}

File: legacy-cpp/src/desk/fee_accrual_engine.cc
#include "tkr/desk/fee_accrual_engine.h"

#include <algorithm>
#include <climits>

namespace tkr {
namespace desk {
namespace {

struct FeeBreakdown {
  std::int64_t commission_cents;
  std::int64_t exchange_cents;
  std::int64_t clearing_cents;
  std::int64_t regulatory_cents;
};

FeeBreakdown ComputeFeeBreakdown(std::int64_t notional_cents,
                                 std::uint32_t qty_milli) {
  FeeBreakdown bd{};
  bd.commission_cents = (notional_cents * 5) / 10000;
  bd.exchange_cents = (notional_cents * 1) / 10000;
  bd.clearing_cents = (static_cast<std::int64_t>(qty_milli) * 2) / 1000;
  bd.regulatory_cents = (notional_cents * 2) / 10000;
  return bd;
}

}  // namespace

FeeAccrualEngine::FeeAccrualEngine(FeeAccrualEngineConfig config)
    : config_(config) {
  LoadDefaultSchedules();
}

void FeeAccrualEngine::LoadDefaultSchedules() {
  FeeSchedule flat{};
  flat.kind = FeeScheduleKind::kFlatPerTrade;
  flat.schedule_id = 1;
  flat.account_id = 0;
  flat.tiers.push_back(FeeTier{0, INT64_MAX, 0, 250});
  schedules_.push_back(flat);

  FeeSchedule tiered{};
  tiered.kind = FeeScheduleKind::kTieredNotional;
  tiered.schedule_id = 2;
  tiered.account_id = 0;
  tiered.tiers.push_back(FeeTier{0, 10000000, 10, 0});
  tiered.tiers.push_back(FeeTier{10000000, 100000000, 5, 0});
  tiered.tiers.push_back(FeeTier{100000000, INT64_MAX, 2, 0});
  schedules_.push_back(tiered);

  FeeSchedule per_share{};
  per_share.kind = FeeScheduleKind::kPerShareMilli;
  per_share.schedule_id = 3;
  per_share.account_id = 0;
  per_share.per_share_milli_cents = 3;
  schedules_.push_back(per_share);
}

void FeeAccrualEngine::RegisterSchedule(const FeeSchedule& schedule) {
  schedules_.push_back(schedule);
}

void FeeAccrualEngine::ClearSchedules() {
  schedules_.clear();
  ledger_.clear();
}

std::int64_t FeeAccrualEngine::AccrueFlatPerTrade(
    const FeeSchedule& schedule) const {
  if (schedule.tiers.empty()) {
    return 0;
  }
  return schedule.tiers[0].flat_fee_cents;
}

std::int64_t FeeAccrualEngine::AccrueTieredNotional(
    const FeeSchedule& schedule, std::int64_t notional_cents) const {
  for (const FeeTier& tier : schedule.tiers) {
    if (notional_cents >= tier.notional_floor_cents &&
        notional_cents < tier.notional_ceiling_cents) {
      const std::int64_t rate_fee =
          (notional_cents * tier.rate_bp) / 10000;
      return rate_fee + tier.flat_fee_cents;
    }
  }
  return 0;
}

std::int64_t FeeAccrualEngine::AccruePerShareMilli(
    const FeeSchedule& schedule, std::uint32_t qty_milli) const {
  return (static_cast<std::int64_t>(qty_milli) *
          schedule.per_share_milli_cents) /
         1000;
}

std::int64_t FeeAccrualEngine::AccrueMonthlyMinimum(
    const FeeSchedule& schedule, std::int64_t running_total) const {
  if (running_total >= schedule.monthly_minimum_cents) {
    return 0;
  }
  return schedule.monthly_minimum_cents - running_total;
}

std::int64_t FeeAccrualEngine::ComputeTradeFee(const FeeSchedule& schedule,
                                               std::uint32_t qty_milli,
                                               std::int64_t notional_cents) const {
  std::int64_t base_fee = 0;
  switch (schedule.kind) {
    case FeeScheduleKind::kFlatPerTrade:
      base_fee = AccrueFlatPerTrade(schedule);
      break;
    case FeeScheduleKind::kTieredNotional:
      base_fee = AccrueTieredNotional(schedule, notional_cents);
      break;
    case FeeScheduleKind::kPerShareMilli:
      base_fee = AccruePerShareMilli(schedule, qty_milli);
      break;
    case FeeScheduleKind::kPerformanceFee:
      base_fee = (notional_cents * 2000) / 10000;
      break;
    case FeeScheduleKind::kMonthlyMinimum:
      base_fee = schedule.monthly_minimum_cents;
      break;
    default:
      break;
  }

  const FeeBreakdown bd = ComputeFeeBreakdown(notional_cents, qty_milli);
  return base_fee + bd.exchange_cents + bd.clearing_cents + bd.regulatory_cents;
}

void FeeAccrualEngine::UpsertLedger(std::uint32_t account_id,
                                    std::uint32_t schedule_id,
                                    std::int64_t fee_cents,
                                    std::int64_t notional_cents) {
  for (AccrualEntry& entry : ledger_) {
    if (entry.account_id == account_id && entry.schedule_id == schedule_id) {
      entry.accrued_cents += fee_cents;
      entry.notional_basis_cents += notional_cents;
      entry.trade_count += 1;
      return;
    }
  }

  AccrualEntry entry{};
  entry.account_id = account_id;
  entry.schedule_id = schedule_id;
  entry.accrued_cents = fee_cents;
  entry.notional_basis_cents = notional_cents;
  entry.trade_count = 1;
  entry.accrual_date_yyyymmdd = config_.accrual_date_yyyymmdd;
  ledger_.push_back(entry);
}

FeeAccrualEngineResult FeeAccrualEngine::AccrueBatch(
    const BatchWireFrame& frame) {
  FeeAccrualEngineResult result{};
  result.status = Status::kOk;

  for (const WireBatchRecord& rec : frame.records) {
    const std::int64_t notional =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;

    for (const FeeSchedule& schedule : schedules_) {
      if (schedule.account_id != 0 &&
          schedule.account_id != rec.account_id) {
        continue;
      }

      const std::int64_t fee =
          ComputeTradeFee(schedule, rec.qty_milli, notional);

      if (schedule.kind == FeeScheduleKind::kMonthlyMinimum) {
        std::int64_t running = 0;
        for (const AccrualEntry& entry : ledger_) {
          if (entry.account_id == rec.account_id &&
              entry.schedule_id == schedule.schedule_id) {
            running = entry.accrued_cents;
            break;
          }
        }
        const std::int64_t min_top_up =
            AccrueMonthlyMinimum(schedule, running + fee);
        UpsertLedger(rec.account_id, schedule.schedule_id, fee + min_top_up,
                     notional);
        result.total_accrued_cents += fee + min_top_up;
      } else {
        UpsertLedger(rec.account_id, schedule.schedule_id, fee, notional);
        result.total_accrued_cents += fee;
      }
      ++result.schedules_applied;
    }
  }

  result.entries = ledger_;
  std::sort(result.entries.begin(), result.entries.end(),
            [](const AccrualEntry& a, const AccrualEntry& b) {
              if (a.account_id != b.account_id) {
                return a.account_id < b.account_id;
              }
              return a.schedule_id < b.schedule_id;
            });

  return result;
}

}  // namespace desk
}  // namespace tkr
